### ISSUE TITLE
Link Tracking Performance

### ISSUE THREAD

**Issue by tyrauber** [ankane/ahoy_email#55]
In running some benchmarks with ahoy_email, it looks like Link Tracking adds over 400% overhead over ActiveMailer.  The feature is important, and some overhead is warranted, but I am wondering if there might be a different approach to parsing the html and augmenting the links.

I ran a bunch of different benchmarks.  All benchmarks are for 100 Emails in development, with perform_deliveries = false. HTML content is generated outside of mail and passed as a body attribute.

**Without ahoy_email**
total: 9413.025117999496, average: 94ms

**With ahoy_email**
- No content, open or click links, (Just ahoy_message creation)
  total: 10273.89198200035, average: 103ms
- Content, no open or click links
  total: 10765.760098998726, average: 108ms
- Content, open and no click links
  total: 11681.741282001894, average: 116ms
- Content, open and click links
  total: 45126.064590996975, average: 451ms
- No Content, No open, Just click links
  total: 46715.007976999914, average: 467ms

Everything else in the gem performs really well, adding marginal overhead.  It were possible to design a different approach to link tracking - perhaps using regex instead of nokogiri - then it could be justified using ahoy_email in every scenario, including sending mailing lists at scale.

In the meantime, if performance is required for a specific mailer, then "track click: false" is required.

**Comment by rainerborene** [comment]
I had the same issue last week and I've been thinking of experiment with the `oga` gem because it could performs better in threaded environments.

```ruby
require 'nokogiri'
require 'oga'
require 'benchmark/ips'

html = <<-HTML
blah blah
<a href="http://google.com">It works</a>
HTML

Benchmark.ips do |x|
  x.report('nokogiri') do
    doc = Nokogiri::HTML(html)
    doc.css('a[href]').each do |node|
      node[:href] = 'blah'
    end
    doc.to_html
  end

  x.report('oga') do
    oga = Oga.parse_html(html)
    oga.css('a[href]').each do |node|
      node.set('href', 'blah')
    end
    oga.to_xml
  end

  x.compare!
end
```

Apparently results are the same but I still need to test how it would perform in a threaded environment like Sidekiq.

```
Warming up --------------------------------------
            nokogiri     1.027k i/100ms
                 oga     1.070k i/100ms
Calculating -------------------------------------
            nokogiri     10.601k (± 2.9%) i/s -     53.404k in   5.041803s
                 oga     10.784k (± 2.8%) i/s -     54.570k in   5.064114s

Comparison:
                 oga:    10784.3 i/s
            nokogiri:    10600.7 i/s - same-ish: difference falls within error
```

**Comment by tyrauber** [comment]
I also noticed jobs failing in Sidekiq. Roughly ~8 per 100.  With two different errors - related to url_for and another with database connection timeouts.  Interestingly enough, when I disabled link tracking those errors seemed to vanish. I'll have to see if it is the same on performance under load.

**Comment by ankane** [comment]
Thanks for reporting. Can you share the code you used to benchmark?  Would be great to use it to test improvements.

**Comment by tyrauber** [comment]
Andrew, I threw together a sample application to help facilitate performance testing:
https://github.com/tyrauber/ahoy_email_performance

Comparison:
tracking click, not open (deliver_now):        0.1 i/s
tracking neither open or click (deliver_now):  0.0 i/s - 1.18x slower
tracking open, not click (deliver_now):        0.0 i/s - 1.44x slower
tracking open and click (deliver_now):         0.0 i/s - 5.65x slower

As you will see, the link tracking adds significant overhead. But ultimately, I think the main barrier to performance is database throughput.  Unless it is possible to separate the email generation (and saving in the database) from the sending, the number of database connections is going to limit the amount of emails you can send per minute. Unfortunately, using ActiveJob and deliver_later doesn't help - in fact, it causes more issues, because concurrent sending will quickly result in database timeouts unless you are careful. (You can see this in the sample app by running "rake ahoy_email:performance[1000,50,true]")  

I would love to do the following:

mail = SomeMailer.notify(email)  # to generate and store the email
mail.deliver_later(wait: 1.hour.from_now)  # to send at a later date

Unfortunately, the deliver job still hits the ahoy_email's process method and results in duplicate message records. I think the only way to get any real email throughput with AhoyEmail enabled is to be able to run one worker to generate a bunch of email in sequence and then have another worker handle just the sending in parallel without having to hit the database again.

Any ideas?

**Comment by tyrauber** [comment]
@rainerborene, Actually, it looks like Nokogiri isn't the slowest part of the link tracking process, but repeatedly calling url_for.  In researching the ActiveJob failures I stumbled upon this thread ('https://github.com/rails/rails/issues/19084') which indicated that not only was url_for prone to failure in a multithreaded environment, but also painfully slow.  Unfortunately the failures are only fixed by upgraded to Ruby 2.3. But I was able to figure out an approach to significantly improve link tracking. Instead of calling url_for for every link, call it once at the class level and then use gsub to modify it.  See (#60)
